3-UsingAPIs/Week1/homework/ex3-rollDice.js

```diff
@@ -10,46 +10,53 @@
 ------------------------------------------------------------------------------*/
 
 // TODO Remove callback and return a promise
-function rollDice(callback) {
-  // Compute a random number of rolls (3-10) that the dice MUST complete
-  const randomRollsToDo = Math.floor(Math.random() * 8) + 3;
-  console.log(`Dice scheduled for ${randomRollsToDo} rolls...`);
-
-  const rollOnce = (roll) => {
-    // Compute a random dice value for the current roll
-    const value = Math.floor(Math.random() * 6) + 1;
-    console.log(`Dice value is now: ${value}`);
-
-    // Use callback to notify that the dice rolled off the table after 6 rolls
-    if (roll > 6) {
-      // TODO replace "error" callback
-      callback(new Error('Oops... Dice rolled off the table.'));
-    }
-
-    // Use callback to communicate the final dice value once finished rolling
-    if (roll === randomRollsToDo) {
-      // TODO replace "success" callback
-      callback(null, value);
-    }
-
-    // Schedule the next roll todo until no more rolls to do
-    if (roll < randomRollsToDo) {
-      setTimeout(() => rollOnce(roll + 1), 500);
-    }
-  };
-
-  // Start the initial roll
-  rollOnce(1);
+function rollDice() {
+  return new Promise((resolve, reject) => {
+    // Compute a random number of rolls (3-10) that the dice MUST complete
+    const randomRollsToDo = Math.floor(Math.random() * 8) + 3;
+    console.log(`Dice scheduled for ${randomRollsToDo} rolls...`);
+
+    const rollOnce = (roll) => {
+      // Compute a random dice value for the current roll
+      const value = Math.floor(Math.random() * 6) + 1;
+      console.log(`Dice value is now: ${value}`);
+
+      // Use callback to notify that the dice rolled off the table after 6 rolls
+      if (roll > 6) {
+        // TODO replace "error" callback
+        reject(new Error('Oops... Dice rolled off the table.'));
+        return;
+      }
+
+      // Use callback to communicate the final dice value once finished rolling
+      if (roll === randomRollsToDo) {
+        // TODO replace "success" callback
+        resolve(null, value);
+      }
+
+      // Schedule the next roll todo until no more rolls to do
+      if (roll < randomRollsToDo) {
+        setTimeout(() => rollOnce(roll + 1), 500);
+      }
+    };
+
+    // Start the initial roll
+    rollOnce(1);
+  });
 }
 
 // TODO Refactor to use promise
-rollDice((error, value) => {
-  if (error !== null) {
-    console.log(error.message);
-  } else {
-    console.log(`Success! Dice settled on ${value}.`);
-  }
-});
+rollDice()
+  .then((value) => console.log(value))
+  .catch((error) => {
+    if (error !== null) {
+      console.log(error.message);
+    }
```

Comment:
in our case, since we always reject with error, it will be never `null` so we can remove this `if`